File: sorting/merge_sort.java
class MergeSort {

    public static void main(String[] args) {

        int[] arr = {12, 13, 42, 32, 54, 86, 21, 8, 6, 4, 5, 7};

        MergeSort mergeSort = new MergeSort();
        mergeSort.sort(arr, 0, arr.length - 1);

        mergeSort.printArray(arr);

    }


    void printArray(int[] arr) {
        for (int i : arr) {
            System.out.print(i + " ");
        }
    }


    void sort(int[] arr, int left, int right) {
        if (left > right) return;

        int mid = left + (right - 1) / 2;

        sort(arr, left, mid);
        sort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }

    void merge(int[] arr, int left, int mid, int right) {

        // find sizes of both sub-arrays to be sorted
        int n1 = mid - left + 1;
        int n2 = right - mid;

        // creating temp arrays
        int[] L = new int[n1];
        int[] R = new int[n2];

        // copy data to temp arrays
        for (int i = 0; i < n1; i++) {
            L[i] = arr[left + i];
        }
        for (int i = 0; i < n2; i++) {
            R[i] = arr[mid + 1 + i];
        }

        // starting to merge
        // initial indexes for first and second sub arrays
        int i = 0, j = 0;

        // initial index for merged sub-array
        int k = left;
        while (i < n1 && j < n2) {
            if (L[i] <= R[i]) {
                arr[k] = L[i];
                i++;
            } else {
                arr[k] = R[j];
                j++;
            }
            k++;
        }

        // adding remaining elements of L[] (if any)
        while (i < n1) {
            arr[k] = L[i];
            i++;
            k++;
        }

        // adding remaining elements of R[] (if any)
        while (j < n1) {
            arr[k] = R[j];
            j++;
            k++;
        }
    }
}

File: sorting/merge_sort.kt
/*
  🧠 How It Works:
	  •	Divide: Recursively split the list in half.
	  •	Conquer: Recursively sort each half.
	  • Combine: Merge the sorted halves into one sorted list.

  Time complexity: O(n log n)
  Space complexity: O(n) (due to additional lists during merge)
*/
fun main() {
    val numbers = mutableListOf(38, 27, 43, 3, 9, 82, 10)
    println("Original: $numbers")
    val sorted = mergeSort(numbers)
    println("Sorted:   $sorted")
}

// ===============================================================================

fun mergeSort(arr: MutableList<Int>): MutableList<Int> {
    // Base case: If array has 0 or 1 element, it's already sorted
    if (arr.size <= 1) return arr

    // Step 1: Split the array into two halves
    val mid = arr.size / 2
    val left = arr.subList(0, mid).toMutableList()
    val right = arr.subList(mid, arr.size).toMutableList()

    // Step 2: Recursively sort both halves
    val sortedLeft = mergeSort(left)
    val sortedRight = mergeSort(right)

    // Step 3: Merge the sorted halves
    return merge(sortedLeft, sortedRight)
}

// Helper function to merge two sorted lists into one sorted list
fun merge(left: MutableList<Int>, right: MutableList<Int>): MutableList<Int> {
    val result = mutableListOf<Int>()

    var i = 0 // pointer for left
    var j = 0 // pointer for right

    // Compare elements and add the smaller one to result
    while (i < left.size && j < right.size) {
        if (left[i] <= right[j]) {
            result.add(left[i])
            i++
        } else {
            result.add(right[j])
            j++
        }
    }

    // Add remaining elements from left (if any)
    while (i < left.size) {
        result.add(left[i])
        i++
    }

    // Add remaining elements from right (if any)
    while (j < right.size) {
        result.add(right[j])
        j++
    }

    return result
}
